### ISSUE TITLE
Importing getFunctions() throws "self is not defined" during SSR/Static (v9-beta)

### ISSUE THREAD

**Issue by MarcusCemes** [firebase/firebase-js-sdk#4846]
### Describe your environment

  * Operating System version: **Ubuntu 20.04.2.0** 
  * Browser version: **Chrome 90.0.4430.93**
  * Firebase SDK version: **9.0.0-beta.1**
  * Firebase Product: **functions**

### Describe the problem

**Unable to render a statically built site using Sveltekit.** This seems to be a result of Sveltekit's new bundling technique of "using Vite to dump everything into one file before running it through Rollup for tree-shaking and optimisation". 

The following line gets included in the bundle when a user's code has `import { getFunctions } from "firebase/functions";`, even if `getFunctions()` is not called. During SSR, `window`, here referenced as `self`, is not available. This is a import side-effect. `fetch()` _is_ pollyfilled during SSR by Sveltekit. From what I can tell, it's the ESM build that's being imported.

```js
registerFunctions(fetch.bind(self))
```

The static build adapter for Sveltekit fails with:

```text
> self is not defined
ReferenceError: self is not defined
    at file:///sandbox/.svelte/output/server/app.js:806:30
    (...)
```

This is not an attempt to use Firebase during SSR or from Node.js. I would like to include firebase in the browser bundle. A solution is lazily importing `getFunctions()` from the browser, however, this comes with a latency penalty and an extra network request.

#### Steps to reproduce:

1. Add `firebase@9.0.0-beta.1` to a Sveltekit application
2. Try to run `npm run dev` or `npm run build` (dev will only fail once, as dev-time SSR is not retried a second time)

[![Edit firebase-self-not-defined](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/firebase-self-not-defined-75k9w?fontsize=14&hidenavigation=1&theme=dark)

The sandbox is a little weird, it occasionally works, occasionally doesn't. **Downloading the sandbox as a ZIP on Windows and Ubuntu still causes the build error.**

##### Offending line

https://github.com/firebase/firebase-js-sdk/blob/9df03c97c31521c29f12a9a3ff2a1d9db625725a/packages-exp/functions-exp/src/index.ts#L30

##### Failure case

```js
import { getFunctions } from "firebase/functions";

export function functions() {
    return getFunctions(app());
}
```

**Comment by MarcusCemes** [comment]
After more testing., even lazy-loading `getFunctions()` via `import("firebase/functions")` from a central file does not solve the issue. My guess is that any import, including those that are used to construct queries such as `doc()` or `collection()`, cause this top-level code to be inlined in the bundle.

The new modular firebase system means these types of on-demand imports are scattered everywhere, using `import("...")` everywhere would incur significant refactoring costs, potentially increase the number of network requests and make the library a lot less tree-shakable.

**Comment by Feiyang1** [comment]
Just want to clarify that the issue happens during static site generation which does happen in Nodejs, so the error makes sense. 

And you're right that it's caused by the side effect of the module, so it happens even if you don't call `getFunctions()`. I'd like to get rid of the side effect eventually, but we need to do some research first which I haven't had a chance to do.

As a workaround for now, is it possible to configure svelte (rollup) to pick up the node build instead of the browser build of `firebase/functions`? To do that, you need to tell rollup to use the `main` field in package.json.

**Comment by MarcusCemes** [comment]
You are correct, the issue does indeed happen during static site generation in a Node.js environment. I understand that firebase expects a browser environment, but SSR requires running some code on the server, before being sent to the browser.

I would argue that **this behaviour does not make sense** and is therefore undesirable. No function from the firebase module is ever used. Import-time initialisation is a bit of a hack from the CommonJS days, introduces start-up delays and makes it impossible to intercept errors. In my mind, there is no technical barrier that would keep initialisation from happening upon the first invocation of a function. I'm not sure how other SSR frameworks (Next.js, Nuxt, ...) have mitigated this issue so far, perhaps Webpack's module system has some form of lazy execute-once-required policy.

While it's true that SvelteKit and Vite are relatively "new kids on the block" that do things differently to battle-tested CommonJS-based frameworks, I commend them on their bravery to try and give the JS community that push into the ESM world that brings it more in line with traditional programming languages. I would like to add that `fetch` _is_ available in the global scope during SvelteKit SSR (library implementation, now in most frameworks), that is not the problem, Node.js just doesn't have a `window` (or `self`) object but instead a `global` (and `this` respectively, incidentally also works in Chrome).

I'm not sure if such a workaround would work, my understanding is that Vite will attempt to transform CJS into ESM, likely resulting in the same issue. Having said that, **I have just discovered a SvelteKit-specific workaround**. Moving the `firebase` dependency to `devDependencies`  will no longer inline `firebase/functions` code in the server bundle, but rather use the ES `import { ... } from ...` syntax. From what I call tell, Node.js does not execute any top-level code (the side effects) when resolving the import, a sort of "runtime tree-shaking", neither does the browser, and the SDK still works.

> In ESM, the module loader runs in asynchronous phases. In the first phase, it parses the script to detect calls to import and export **without running the imported script**. ([Source](https://redfin.engineering/node-modules-at-war-why-commonjs-and-es-modules-cant-get-along-9617135eeca1))

Heads up, I'm not sure how browsers handle this case, but if ESM hits more mainstream frameworks, it may become a problem. For now, my workaround is adequate, although I would like the new firebase SDK to embrace true ESM modularity.

**Comment by benmccann** [comment]
> As a workaround for now, is it possible to configure svelte (rollup) to pick up the node build instead of the browser build of firebase/functions? 

If I'm looking at the right package source (I'm on my phone and may not be), then I see the following:

>  "main": "dist/index.node.cjs.js",
  "browser": "dist/index.esm.js",
  "module": "dist/index.esm.js",

This seems like it may not be setup correctly. `module` is something used by Node.js and thus implies server side. If `index.esm.js` is isomorphic that's fine, but if it's intended to be used only on the browser I think that's wrong and the field should either be removed from `package.json` or an ESM version of the server side build can be bundled and listed there

**Comment by Feiyang1** [comment]
@benmccann Interesting! We have been using the module field for browser builds all the time. I just checked rollup documentation again, but didn't see any text that says whether it should be used for nodejs build or browser build. so my understanding is that there is no consensus on that, and it's up to the lib authors. Do you have any link that says module field should point to nodejs build?

**Comment by benmccann** [comment]
`module` isn't quite standardized and is more of a de facto standard, so isn't covered in the official npm docs. However, `main` has always been the Node entry point with `browser` being the client-side entry point. [These are the two standardized fields explicitly covered in the documentation.](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#main). The point of `module` is to differentiate ESM entry point from CommonJS entry point as discussed in [the Node docs](https://nodejs.org/api/packages.html#packages_dual_commonjs_es_module_packages) and [Rollup docs](https://github.com/rollup/rollup/wiki/pkg.module). In both these places `module` is explicitly being discussed as an alternative to `main`, which is the server-side entry point (with the client entry point being `browser`).

**Comment by benmccann** [comment]
Node only knows about `main`. Webpack, Rollup, etc. use `module`. By default, they both consider `module` for both server and client, but only consider `browser` for client. So I do think that `module` should be able to run on the server and that `browser` should be used for the client.

I also think there's also a Vite bug here. I can fix the codesandbox example with:
```
		vite: {
			ssr: {
				noExternal: ['@firebase/app', '@firebase/functions']
			}
		}
```

That contradicts [the behavior described in the Vite docs](https://vitejs.dev/guide/ssr.html#ssr-externals), which says that if `module` and `main` are different it will use `main`. Here, that's not whats happening. It's using `module`, which is causing it to fail.

So I think that both Vite and Firebase should be changed. If we change either one that should allow Firebase to work with Vite, but it seems to me that neither is currently configured the way I'd expect

**Comment by benmccann** [comment]
It looks like the codesandbox example is using an outdated version of SvelteKit and Vite. If I update it, then Vite's behavior does match that described in the docs and it does use the CJS version on the server-side. Hurray for that! It then gives a new error `exports is not defined`, which is the #1 issue in the Vite tracker: https://github.com/vitejs/vite/issues/2579

**Comment by benmccann** [comment]
Okay, digging into Vite + Firebase some more, it seems that `tryNodeResolve` fails on Firebase. Not sure why...

**Comment by benmccann** [comment]
It turns out that code path shouldn't be executed in Vite and there's actually an underlying bug. Here's a bug I just filed in the Vite tracker for the Vite side of this: https://github.com/vitejs/vite/issues/4425

We should still leave this open though as I still think it's wrong to put a browser-only build in `module`

**Comment by MarcusCemes** [comment]
I've re-created the sandbox using a new Sveltekit skeleton project (so removed the explicit vite dependency), updated all the dependencies to their respective latest versions (beta/next) and locked in the versions to maintain reproducibility. The bug is still there, you just need to **download as a ZIP** and run it locally, I think it's a caching issue with codesandbox, Ctrl+F5 usually helps to trigger the error.

I have managed to get this to work on my own project through extensive dynamic `import(...)` and `if (browser)` checks, but it would be preferable to be able to import the cloud functions module without side-effects and only instantiate it once it is loaded (once it is needed) in the browser, without additional code-splitting head-aches.

[![Edit firebase-self-not-defined (2)](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/firebase-self-not-defined-2-bbdst?fontsize=14&hidenavigation=1&theme=dark)

<details>
<summary>Terminal error</summary>


```text
ReferenceError: self is not defined
    at eval (/node_modules/@firebase/functions/dist/index.esm2017.js?v=ecda2b99:651:30)
    at instantiateModule (C:\Users\Marcus\dev\experimental\sandbox\node_modules\vite\dist\node\chunks\dep-c1a9de64.js:73464:166)
```

</details>

**Comment by rawkode** [comment]
@MarcusCemes I'm hitting this error and I'd love to work around it as you have. Would you be able to share some code to show how to protect against the self is not defined error, please?

**Comment by rawkode** [comment]
Found something that works, for anyone else that runs into this:


```
<script lang="ts">
	import { browser } from '$app/env';

	const doIt = () => {
		if (browser) {
			import('firebase/functions').then((fns) => {
				const functions = fns.getFunctions();

                                // ... your code here
			});
		} else {
			console.log('Not in browser');
		}
	};
</script>
```

**Comment by MarcusCemes** [comment]
@rawkode At the moment your best bet is to lazily import firebase using `import(...)`, like in your example above. Vite will separate out the library into a different file that can (and must) be fetched and loaded via an extra network request when the `import(...)` function is run, the downside is the slightly higher latency and error handling, but it's the best I've got. Of course, moving the problem to runtime requires you to take care not to accidentally call it during SSR.

For my application, I have "core" files that offer an abstraction over firebase and import using the ES syntax, "glue" files *such as queries and stores) that lazy-`import(...)` "core" that I can import directly using the ES syntax from my application (either in `onMount` or as a result of a DOM event, such as a click, otherwise it may run during SSR). In my mind, this lets Vite efficiently bundle all firebase library code, and I have cleaner ES imports in my main application code with Intellisense (and I can do some additional sanitizing between my application and firebase).

Cloud Functions are a little easier, I just created my own async abstraction that loads the library and executes the function, it just has an extra initial network request.

```ts
// app.svelte
import { runFunction } from "$lib/functions.ts";

async function handleEvent() {
  const result = await runFunction(Functions.Ingest, { ... });
}

// functions.ts
async function runFunction<P, R>(name: Functions, params: P): Promise<R> {
  const { cloudFunction } = await import("./core/functions"); // this lazy import is important
  const fn = await cloudFunction<P, R>(name);
  return (await fn(params)).data;
}

// core/functions.ts
import { getFunctions, httpsCallable } from "firebase/functions";
import { firebase } from "./firebase.ts";

export async function cloudFunction<T, R>(name: string): Promise<HttpsCallable<T, R>> {
  const fns = getFunctions(firebase(), "europe-west3");
  return wrapCloudFunction(httpsCallable(fns, name)); // sanitize payload, remove undefined values, cause error
}
```

I'm not sure if importing Firestore is safe, I've spent so long fighting SSR errors that I've moved everything to lazy `import(...)`s, just to be on the safe side. Svelte stores offer a pretty neat abstraction to Firestore snapshot queries!

Now here's the problem, when a Svelte component unmounts, any store subscriptions are cancelled (by calling the given unsubscribe function), this is important when subscribing to a Firestore snapshot otherwise it would leave dangling WebSockets and charge you unnecessary reads! The act of lazy-importing "firebase/firestore" means we can't immediately return an unsubscribe function (synchronous), as we're in an asynchronous context...

As a result, I made a few extra helpers using browser APIs (abort signals) to guarantee clean-up. Yes, it's a bit of a monstrosity, but it works, it's (fairly) nicely typed, it shouldn't be necessary in the future. My experiments show that Svelte stores are eagerly evaluated 

```ts
// app.svelte

  const store = filmsStore();
import { readable } from "svelte/store";
import type { Readable } from "svelte/store";

export type ApiData<T> =
  | { state: "pending" }
  | { state: "error"; error: Error }
  | { state: "ok"; value: T };

export function filmStore(filmId?: string): Readable<ApiData<Film | null>> {
  if (!browser || !filmId) return readableValue(apiPending()); // this is important
  return listenTo(async () => {
    const { filmQuery } = await import("./queries");
    return filmQuery(filmId);
  });
}

type Mapped<T extends DocumentReference | Query> = T extends DocumentReference<infer U>
  ? U
  : T extends Query<infer U>
  ? Id<U>[]
  : never;

export function listenTo<T extends DocumentReference | Query>(
  queryFn: () => Promisable<T>,
): Readable<ApiData<Mapped<T>>> {
  return asyncReadable(apiPending(), async (set, signal) => {
    const query = await queryFn();
    if (signal.aborted) return;
    return await createSnapshot(query, set);
  });
}

function asyncReadable<T>(
  init: T,
  subscriber: (set: Subscriber<T>, signal: AbortSignal) => Promisable<(() => void) | void>,
): Readable<T> {
  return readable(init, (set) => {
    const ctrl = new AbortController();
    const signal = ctrl.signal;

    void (async () => {
      const result = await subscriber(set, signal);

      if (typeof result === "function") {
        if (signal.aborted) {
          result();
        } else {
          signal.addEventListener("abort", result);
        }
      }
    })();

    return () => ctrl.abort();
  });
}


async function createSnapshot<T extends DocumentReference | Query>(
  query: T,
  set: Subscriber<ApiData<Mapped<T>>>,
): Promise<Unsubscribe> {
  const { onSnapshot, DocumentReference, Query } = await import("firebase/firestore");
  if (query instanceof DocumentReference) {
    return onSnapshot(
      query,
      (data) => set(apiData(data.data() as Mapped<T>)),
      (error) => set(apiError(error)),
    );
  } else if (query instanceof Query) {
    return onSnapshot(
      query,
      (data) =>
        set(apiData(data.docs.map((doc) => ({ id: doc.id, data: doc.data() })) as Mapped<T>)),
      (error) => set(apiError(error)),
    );
  }
}

// Store that emits a value once
function readableValue<T>(value: T): Readable<T> {
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  return readable(value, () => {});
}
```

**Comment by Xstyler85** [comment]
I have the same problem:
```
ReferenceError: self is not defined
    at eval (/node_modules/@firebase/functions/dist/index.esm2017.js?v=093f313f:651:30)
    ...
```
I'm looking forward to an easier workaround / fix.

**Comment by Feiyang1** [comment]
@MarcusCemes IIUC, Vite translates `import 'firebase/functions'` to `require('firebase/functions')` during SSR, which should pick up the node version of functions correctly. Is it not the case?

**Comment by MarcusCemes** [comment]
@Feiyang1 I am not too familiar with the inner workings of the Vite/Rollup/Sveltekit trio, but inspecting the generated `server/app.js` file for SSR, the ESM version of `firebase/functions` seems to have been _inlined_ into the file, like so, and not converted into CommonJS `require(...)` imports:

```js
const name = "@firebase/functions-exp";
const version = "0.0.900-exp.8b4d7550f";
registerFunctions(fetch.bind(self));      // offending line
registerVersion(name, version);
const Test = create_ssr_component(($$result, $$props, $$bindings, slots) => {
```

As stated in my above comment, moving `firebase` to `devDependencies` does create an explicit `import ... from` in `server/app.js` for some reason instead of inlining, this does solve the issue, but **only** because the Node.js ESM loader does not execute any top-level side-effect code! The offending lines never run, and the behaviour may be the same in the browser using the ESM bundle, I'm not sure.

The `firebase/functions`'s `package.json` advertises its bundles like so:

```json
{
  "name": "firebase-exp/functions",
  "main": "dist/index.cjs",
  "browser": "dist/index.esm.js",
  "module": "dist/index.esm.js",
  "typings": "dist/functions/index.d.ts",
  "type": "module"
}
```

My guess is that Vite is inlining the `module` bundle (which happens to be the browser bundle as well) as it is an ESM first bundler. Technically, ESM _should_ work in a Node.js environment, it's purely an import mechanism and does not require any browser-specific APIs.

**Comment by benmccann** [comment]
@Feiyang1 if you'd like to continue using the `module` field for browser builds all the time, the other way to fix this would be to add `exports` to the `package.json` to clarify which builds should be used by Node vs the browser

**Comment by Bandit** [comment]
FWIW I encounter this (and this https://github.com/sveltejs/kit/issues/612) issue when switching from `import { ... } from 'firebase/firestore'` to `import {...} from 'firebase/firestore/lite'`

**Comment by MarcusCemes** [comment]
To follow up on the discussion, here's my experience with SvelteKit/Vite/Firebase, for anyone else who might find this insightful.

There's a lot going on with the Firebase SDK that Vite and Node.js can't deal with (without resorting to the Admin SDK). I've encountered all sorts of errors after accidentally including a Firebase import somewhere in the dependency chain, whether on purpose or by accident with VSCode's Intellisense. I have a spaghetti of browser checks and lazy `import`s, and it's still not working correctly.

I tried moving SSR-related Firestore code into a Cloud Function that I can query during SSR without the SDK, before realising that `Timestamp`s and `DocumentReference`s are not trivially serialized over JSON (and it plays havoc with types...). Convenient, but locked-in.

That, and also [this](https://issuetracker.google.com/issues/158014637) (5-10 seconds for me...), and [this](https://github.com/firebase/firebase-tools/issues/3751), has led me to start the move away from Firebase. I was very excited at first, but it's quite difficult to integrate these stacks together at the moment, without too many unnecessary abstractions. I've no doubt that for some it may be a fantastic solution.

The client SDK is used by a lot more than just SvelteKit/Vite users, it was designed with CommonJS and Webpack in mind, I'm guessing that messing with the `package.json` could break things for someone else, even if it is more correct. I'm sure the Google engineers are doing their best, and I have the utmost respect for their work, but for my use case, it was simply too much hassle to be able to get anything done, I miss the days of a simple JSON API and concise hand-written code.
